lnbits/lnbits
# Paid extensions

### Summary
Add support for paid extensions.
It is now possible for extension developers to require payment for their extension.
When an admin user wants to install an extension, it will be prompted with a request for payment. The payment can be made:
 - from one of the user's LNbits wallets, or
 - from an external wallet

https://github.com/lnbits/lnbits/assets/2951406/bb9ae8ea-2468-43d2-8c1f-84594e825e92

**Features**:
 - paid releases are required to have a `pay_link` field, see description [here](https://github.com/lnbits/lnbits-extensions/blob/main/README.md#paid-extensions). 
    - multiple releases can have the same `pay_link` (no need to pay again for a new release install)
    - but it is possible that some releases (eg. major upgrade) have their own `pay_link`, thus requiring a new payment to install them
- if an extension is un-installed then the payment history is lost

**Serve the extension file**:
 - install  paywall extension version 0.3.0 or higher. Check [this video](https://github.com/lnbits/paywall/blob/main/README.md#file-paywall) for creating a paywall for the extension zip file
 - or use this extensions.json test file in your server configuration
   - `https://raw.githubusercontent.com/lnbits/lnbits-extensions/paid_extensions/extensions.json`
   - the first 3 releases have a cost of `2 sats` (basic features), the next 3 have a cost of `5 sats` (premium)

**Test Scenarios**
Prerequisites: configure the repo for the paid extension (see above description).
 - install the extension by paying with a LNbits wallet
   - test also a wallet with no funds (should fail)
 - install the extension by displaying the QR code and paying with an external wallet
 - install the extension by displaying the QR code
    - after the QR code is displayed copy the invoice, but do not pay
    - refresh the page
    - pay the invoice
    - go back to the install dialog, and click `Retry Install`
 - uninstall the extension -> the payment history will be lost
